rust-runtime/aws-smithy-http-server/src/routing/mod.rs

```diff
@@ -150,42 +313,84 @@ where
     fn call(&mut self, req: Request<B>) -> Self::Future {
         let mut method_not_allowed = false;
 
-        for (route, request_spec) in &self.routes {
-            match request_spec.matches(&req) {
-                request_spec::Match::Yes => {
-                    return RouterFuture::from_oneshot(route.clone().oneshot(req));
+        match &self.routes {
+            // REST routes.
+            Routes::RestJson1(routes) | Routes::RestXml(routes) => {
+                // Loop through all the routes and validate if any of them matches. Routes are already ranked.
+                for (route, request_spec) in routes {
+                    match request_spec.matches(&req) {
+                        request_spec::Match::Yes => {
+                            // We have a match, return the Router future.
+                            return RouterFuture::from_oneshot(route.clone().oneshot(req));
+                        }
+                        request_spec::Match::MethodNotAllowed => method_not_allowed = true,
+                        // Continue looping to see if another route matches.
+                        request_spec::Match::No => continue,
+                    }
+                }
+
+                if method_not_allowed {
+                    // The HTTP method is not correct.
+                    self.method_not_allowed()
+                } else {
+                    // In any other case return the `RuntimeError::UnknownOperation`.
+                    self.unknown_operation()
                 }
-                request_spec::Match::MethodNotAllowed => method_not_allowed = true,
-                // Continue looping to see if another route matches.
-                request_spec::Match::No => continue,
+            }
+            // AwsJson routes.
+            Routes::AwsJson10(routes) | Routes::AwsJson11(routes) => {
+                if req.uri() == "/" {
+                    // Check the request method for POST.
+                    if req.method() == http::Method::POST {
+                        // Find the `x-amz-target` header.
+                        if let Some(target) = req.headers().get("x-amz-target") {
+                            if let Ok(target) = target.to_str() {
+                                // Search inside the HashMap for a route for the target.
+                                let route = routes.get(target);
+                                if let Some(route) = route {
+                                    // We have a match, return the Router future.
+                                    return RouterFuture::from_oneshot(route.clone().oneshot(req));
+                                }
+                            }
+                        }
+                    } else {
+                        // The HTTP method is not POST.
```

Comment:
This will return `405 Method Not Allowed` when we don't know if the target resource (indicated by the `x-amz-target` header) exists.

I also identified another problem according to the HTTP spec: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405

> The server must generate an Allow header field in a 405 status code response. The field must contain a list of methods that the target resource currently supports.

---

Maybe the AWS protocols don't use `Method Not Allowed` at all, it's not in their spec. What does `smithy-typescript` do?

Reply:
AFAIK this is the first language implementing awsJson spec, I was not able to find any reference of awsJson inside the smithy-typescript repo.